File: frontend/src/types/snapshot.ts
export interface GraphNode {
  id: string;
  label?: string;
  [key: string]: unknown;
}

export interface GraphEdge {
  source: string;
  target: string;
  weight?: number;
  [key: string]: unknown;
}

export interface SnapshotState {
  current_node?: string | null;
  current_edge?: string | null;
  visited?: string[];
  queue?: string[];
  distances?: Record<string, number>;
  traversed_edges?: string[];    // Các cạnh đã đi qua (tô màu xanh dương)
  final_path_edges?: string[];   // Kết quả cuối (MST / đường đi ngắn nhất) — tô màu xanh lá
  mst_edges?: string[];
  selected_edges?: string[];
  notes?: string;
  action_description?: string;
  [key: string]: unknown;
}

export interface AlgorithmSnapshot {
  step_id: number;
  algorithm: string;
  graph: {
    nodes: GraphNode[];
    edges: GraphEdge[];
    [key: string]: unknown;
  };
  state: SnapshotState;
  meta?: Record<string, unknown>;
}

/**
 * Response thực tế từ backend GET /api/v1/step/{session_id}
 * Backend trả về step_data_json dạng flat, không có graph
 */
export interface StepResponse {
  session_id: string;
  step_index: number;
  description: string;
  data: {
    phase_id: string;
    current_node: string | null;
    target_node: string | null;
    visited: string[];
    distances: Record<string, number>;
    queue: string[];
  };
}

export interface InitSessionResponse {
  session_id: string;
  total_steps: number;
  algorithm: string;
}

export interface RagExplanationResponse {
  explanation: string;
  session_id: string;
  step_index: number;
}

File: backend/app/schemas/snapshot.py
from typing import Any

from pydantic import BaseModel, Field


class GraphNode(BaseModel):
    id: str
    label: str | None = None
    extra: dict[str, Any] = Field(default_factory=dict)


class GraphEdge(BaseModel):
    source: str
    target: str
    weight: float | None = None
    extra: dict[str, Any] = Field(default_factory=dict)


class GraphPayload(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]


class SnapshotState(BaseModel):
    current_node: str | None = None
    visited: list[str] = Field(default_factory=list)
    queue: list[str] = Field(default_factory=list)
    distances: dict[str, float] = Field(default_factory=dict)
    notes: str | None = None


class AlgorithmSnapshot(BaseModel):
    step_id: int = Field(ge=0)
    algorithm: str
    graph: GraphPayload
    state: SnapshotState
    meta: dict[str, Any] = Field(default_factory=dict)
